Added Pokemon-style Narrative Display

102109190 opened the pull request:
### Changes Made

- Added three scripts (DialogueBehaviour.cs, Dialogue.cs & DialogueTrigger.cs)
- Added DialogueDisplay prefab to Canvas in Gameplay scene

### How to test

- Open up Gameplay scene and navigate past MenuNarrative to see DialogueDisplay in action

102113528 commented on MicrowaveGame/Assets/Resources/Scripts/Dialogue/Dialogue.cs:
```diff
@@ -0,0 +1,18 @@
+using UnityEngine;
+
+[System.Serializable]
+public class Dialogue
```
This should be in the namespace `namespace Scripts.Dialogue { ... }`.

102113528 replied:
This should be in the namespace `namespace Scripts.Dialogue { ... }`.

102113528 commented on MicrowaveGame/Assets/Resources/Scripts/Dialogue/DialogueBehaviour.cs:
```diff
@@ -0,0 +1,86 @@
+    /// </summary>
+    /// <param name="sentence">Line of dialogue to be typed out</param>
+    /// <returns></returns>
+    IEnumerator TypeSentence (string sentence)
+    {
+        ContentText.text = "";
+
+        foreach (char letter in sentence.ToCharArray())
+        {
+            ContentText.text += letter;
+            yield return null;
+        }
+    }
```
This should use some form of fixed update / timer so characters are added at a consistent rate (i.e. currently fast computers zoom through the animation).

102113528 commented on MicrowaveGame/Assets/Resources/Scripts/Dialogue/DialogueBehaviour.cs:
```diff
@@ -0,0 +1,86 @@
+    /// Sets up the Dialogue object for display
+    /// </summary>
+    /// <param name="dialogue">Dialogue object with all dialogue strings</param>
+    public void StartDialogue(Dialogue dialogue)
+    {
+        Time.timeScale = 0.0f;
+
+        SpeakerName.text = dialogue.Name;
+
+        Sentences.Clear();
+
+        foreach (string sentence in dialogue.Sentences)
+        {
+            Sentences.Enqueue(sentence);
+        }
+
+        DisplayNextSentence();
+    }
```
You should add the following line so the input actions are disabled when starting the dialogue.
```csharp
GameObject.Find("Player").GetComponent<PlayerInput>().actions.Disable();
```

102113528 commented on MicrowaveGame/Assets/Resources/Scripts/Dialogue/DialogueBehaviour.cs:
```diff
@@ -0,0 +1,86 @@
+    /// <summary>
+    /// Text objects where strings are entered into 
+    /// </summary>
+    private Text SpeakerName; 
+    private Text ContentText;
```
Private fields use `_camelCase` (i.e. `_speakerName`, `_contentText`).

102113528 commented on MicrowaveGame/Assets/Resources/Scripts/Dialogue/DialogueBehaviour.cs:
```diff
@@ -0,0 +1,86 @@
+    /// <summary>
+    /// Queue that strings are loaded into.
+    /// </summary>
+    private Queue<string> Sentences;
```
Private fields use `_camelCase` (i.e. `_sentences`).

102113528 commented on MicrowaveGame/Assets/Resources/Scripts/Dialogue/DialogueBehaviour.cs:
```diff
@@ -0,0 +1,86 @@
+
+    public void Awake()
+    {
+        Sentences = new Queue<string>(); // Saw a circularQueue script in Utilities, can switch to that if preferred
```
Queue is correct for this case. No need for CircularQueue.

102113528 commented on MicrowaveGame/Assets/Resources/Scripts/Dialogue/DialogueBehaviour.cs:
```diff
@@ -0,0 +1,86 @@
+    /// <summary>
+    /// Ends the dialogue, and de-activates the gameObject
+    /// </summary>
+    void EndDialogue()
+    {
+        Time.timeScale = 1.0f;
+        gameObject.SetActive(false);
+    }
```
You should add the following line so the input actions are enabled once the dialogue is finished.
```csharp
GameObject.Find("Player").GetComponent<PlayerInput>().actions.Enable();
```

102113528 commented on MicrowaveGame/Assets/Resources/Scripts/Dialogue/DialogueTrigger.cs:
```diff
@@ -0,0 +1,24 @@
+    /// <summary>
+    /// Dialogue Object
+    /// </summary>
+    public Dialogue dialogue;
```
Public fields use `PascalCase` (i.e. `Dialogue`).

102109190 replied:
Would naming the variable the same as the object not break the script (like it wouldn't tell when I'm invoking the object script versus that particular instance?)

102113528 replied:
This would be an issue if you had a field named `Dialogue` in the `Dialogue` class, but since it's inside `DialogueTrigger`, it's okay.

102113528 commented on MicrowaveGame/Assets/Resources/Scripts/Dialogue/DialogueBehaviour.cs:
```diff
@@ -0,0 +1,86 @@
+using UnityEngine;
+using UnityEngine.UI;
+
+public class DialogueBehaviour : MonoBehaviour
```
This should be in the namespace `namespace Scripts.Dialogue { ... }`.

102109190 replied:
implementing the namespace created an issue where the code wants "namespace.DialogueTrigger" when using GetComponent<>. Is that supposed to happen, because I haven't seen that syntax in other scripts from what I've seen

102113528 replied:
Hmmm, there shouldn't be an issue if all three dialogue classes (`Dialogue`, `DialogueBehaviour`, and `DialogueTrigger`) are in the same namespace (`namespace Scripts.Dialogue`).

102109190 replied:
Seems I had to include a "using Scripts.DialogueUI" line. interesting

102113528 replied:
For future reference, try to match the namespace to the folder structure. In this case, that would be `Scripts.Dialogue`.

102113528 commented on MicrowaveGame/Assets/Resources/Scripts/Dialogue/DialogueTrigger.cs:
```diff
@@ -0,0 +1,24 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class DialogueTrigger : MonoBehaviour
```
This should be in the namespace `namespace Scripts.Dialogue { ... }`.

102113528 commented on MicrowaveGame/Assets/Resources/Scripts/Dialogue/DialogueTrigger.cs:
```diff
@@ -0,0 +1,24 @@
+    public void TriggerDialogue()
+    {
+        GameObject.Find("Canvas").transform.Find("DialogueDisplay").gameObject.SetActive(true);
+        FindObjectOfType<DialogueBehaviour>().StartDialogue(dialogue);
```
`FindObjectOfType()` will only find the first DialogueBehaviour it comes across. We should instead pass in an additional field through the Unity inspector that references the object containing the DialogBehaviour on it.

102113528 reviewed (changes requested):
I'm happy to make these changes myself if you're alright with that.

102109190 commented:
### Changes made

- Private variables changed to _camelCase
- Public variables changed to PascalCase
- Changed "yield return null" to "yield return new WaitForSeconds()"
- Changed Dialogue folder to DialogueUI (caused issues with Dialogue class)
- Added namespace references to scripts inside DialogueUI
- Added lines to disable player input in DialogueBehaviour
- Added DialogueBehaviour reference to DialogueTrigger

### How to Test
- Start GameplayScene, go through MenuNarrative to see DialogueDisplay in action
- Review scripts in DialogueUI folder

102113528 commented:
I've refactored the code to use the `Lerped` utility class from Andrew's minimap branch (has been merged into main).
Someone else needs to review / merge this in.

102107466 commented:
Looks good
